modules/packages/LinearAlgebra.chpl

```diff
@@ -605,6 +604,60 @@ private inline proc _raw(D: domain(1), i) {
 // Matrix Structure
 //
 
+
+/*
+   Return a Vector containing the diagonal elements of ``A`` if the argument ``A`` is of rank 2.
+   Return a diagonal Matrix whose diagonal contains elements of ``A`` if argument ``A`` is of rank 1.
+ */
+proc diag(A: [?Adom] ?eltType){
+  if(Adom.rank == 2) then{
+    if(!isSquare(A)) then halt("diag supports only square matrices");
```

Comment:
I don't think we need to impose a square constraint; it would be fine for this function to return ``A[i,i]`` even if the matrix isn't square. 

In the case where the diagonal is passed it, I think returning a square matrix is fine.